### PR TITLE
[14.0][ADD] product_category_inter_company module

### PR DESCRIPTION
**Pull request by Kev-Roche**
This module allows to set product categories as company dependent.

### CODE REVIEW

**Comment by sebastienbeau** [comment]
@Kev-Roche you should inherit the product.template and 

- add a "check_company" on the field categ_id
- add _check_company_auto=True on the class product.template

You can grep example on odoo code

**Comment by Kev-Roche** [comment]
@sebastienbeau done

**Review by flotho — APPROVED** [review]
code review LGTM

File: product_category_inter_company/models/product_category.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
+# Copyright 2022 Akretion (https://www.akretion.com).
+# @author Kévin Roche <kevin.roche@akretion.com>
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import fields, models
+
+
+class ProductCategory(models.Model):
+    _inherit = "product.category"
+
```
Comment (sebastienbeau) [inline product_category_inter_company/models/product_category.py:12]: _check_company_auto = True is missing here

File: product_category_inter_company/views/product_category.xml [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+     @author Kévin Roche <kevin.roche@akretion.com>
+     License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl). -->
+<odoo>
+    <record id="product_category_form_view" model="ir.ui.view">
+        <field name="model">product.category</field>
+        <field name="inherit_id" ref="product.product_category_form_view" />
+        <field name="arch" type="xml">
+            <field name="parent_id" position="after">
+                <field name="company_id" />
+            </field>
```
Comment (sebastienbeau) [inline product_category_inter_company/views/product_category.xml:12]: add a domain on parent_id ['|', ('company_id', '=', False), ('company_id', '=', company_id)]

**Comment by Kev-Roche** [comment]
@sebastienbeau, I added your suggestions.

File: product_category_inter_company/models/product_category.py [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import fields, models
+
+
+class ProductCategory(models.Model):
+    _inherit = "product.category"
+    _check_company_auto = True
+
+    company_id = fields.Many2one(comodel_name="res.company", string="Company")
```
Comment (sebastienbeau) [inline product_category_inter_company/models/product_category.py]: add index=True

File: product_category_inter_company/models/product_category.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+    parent_id = fields.Many2one(check_company=True)
```
Comment (sebastienbeau) [inline product_category_inter_company/models/product_category.py:18]: We can add a python constraint to force all child of a category that belong to a company to belong to this company too.

For example it should be not possible to have
categ A > categ B > categ C

with
Categ A without company
Categ B belong to company 1
Categ C without company

Categ C must belong to company 1 or Categ B must be without company


Something like

```

@api.constraint("parent_id", "company_id")
def check_company_restriction(self):
    for record in self:
        if record.parent_id.company_id and not record.parent_id.company_id != record.company_id:
            raise UserError("The parent category have a company so your category %s must belong to the same company" % record.name)
        if record.company_id:
            for child in records.child_ids:
                if record.company_id != child.company_id:
                    if child.company_id:
                        msg = _("The category %s must be shared as the child %s belong to company %s") % (record.name, child.name)
                    else:
                        msg =  _("The category %s must be shared as the child %s is shared") % (record.name, child.name)
                    raise UserError(msg)

**Review by sebastienbeau — CHANGES_REQUESTED** [review]

**Review by sebastienbeau — APPROVED** [review]

**Comment by sebastienbeau** [comment]
@cvinh it's possible for you to do a review, as you have the same need? thanks

**Pull request merged**